mesonbuild/dependencies/base.py

```diff
@@ -467,6 +473,22 @@ def get_partial_dependency(self, *, compile_args: bool = False,
         return new
 
 
+def get_leaf_external_dependencies(deps: T.List[Dependency]) -> T.List[Dependency]:
+    if not deps:
+        # Ensure that we always return a new instance
+        return deps.copy()
+    final_deps = []
+    while deps:
+        next_deps = []
+        for d in mesonlib.listify(deps):
+            if not isinstance(d, Dependency) or d.is_built():
+                raise DependencyException('Dependencies must be external dependencies')
+            final_deps.append(d)
+            next_deps.extend(d.ext_deps)
+        deps = next_deps
+    return final_deps
```

Comment:
That looks nice, however note that all of the uses I could find start from a _list_ of dependencies. So you would need something like `itertools.chain.from_iterable((iter_deps(d) for d in deps))` which is not pretty either.

Reply:
Another update, I jotted something that takes a list of dependencies and returns a generator of dependencies, but it cannot really be used as much as I hoped:
* it can be used in BuildTarget.add_deps
* it cannot be used in HotdocTargetBuilder.process_dependencies because it accepts pretty much anything and not just dependencies
* it cannot be used very well in GnomeModule._get_dependencies_flags because it recurses on build.SharedLibrary's `lib.get_external_deps()` in addition to dependencies. I made `iter_deps()` could optionally iterate on the external dependencies of an InternalDependency's `dep.libraries` (but I don't understand why not do so with static libraries).
* the pkgconfig module's DependencyHelper._process_libs is also a complicated beast doing mutual recursion with DependencyHelper's other methods _add_lib_dependencies, add_pub_libs and add_priv_libs. I didn't touch it.

All in all, adding iter_deps() ended up adding 6 lines of code. The only file where the lines are reduced is mesonbuild/build.py.